src/replication.c

```diff
@@ -1112,15 +1113,65 @@ void restartAOFAfterSYNC() {
     }
 }
 
+long long getMonotonicTimeMs() {
+    struct timespec res;
+    clock_gettime(CLOCK_MONOTONIC, &res);
+    return (res.tv_sec * 1000000000LL + res.tv_nsec) / 1000000LL;
+}
+
+#define SLAVE_SYNC_BUDGET_UPDATE_INTERVAL_MS 10
+#define SLAVE_SYNC_BUDGET_MAX_INTERVAL_MS 50
+void updateSlaveSyncBudget(ssize_t nread) {
+    if (server.slave_sync_speed_bps <= 0) return;
```

Comment:
May be useful to add comments such as `/* No budget limit */`, `/* Subtract the amount of data transmitted from the budget */`, `/* Add to the budget based on the configured bits per second and the elapsed time */` to document sections of code